### PR TITLE
#397: Support bitvector rotation in formula visitor

### PR DESCRIPTION
**Pull request by kfriedberger**

### CODE REVIEW

File: src/org/sosy_lab/java_smt/test/RotationVisitorTest.java [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+
+    a = bvmgr.makeVariable(8, "a");
+    x = bvmgr.makeBitvector(8, 7);
+    y = bvmgr.makeBitvector(8, 1);
+  }
+
+  private void assertEqualTransformation(BitvectorFormula underTest)
+      throws SolverException, InterruptedException {
+    BooleanFormula f = bvmgr.equal(a, underTest);
+    // System.out.println(mgr.dumpFormula(f)); // TODO dumpFormula crashes with Princess
```
Comment (kfriedberger) [inline src/org/sosy_lab/java_smt/test/RotationVisitorTest.java:41]: We should report this to Princess.

**Review by daniel-raffler — APPROVED** [review]
I've just checked and this seems to take care of the issue in https://github.com/sosy-lab/java-smt/pull/396

LGTM!

File: src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaFormulaCreator.java [hunk 2]
```diff
@@ -305,8 +305,16 @@ private FunctionDeclarationKind getDeclarationKind(Term term) {
...
+      return FunctionDeclarationKind.BV_ROTATE_LEFT;
+    } else if (kind.equals(Kind.BV_ROR)) {
+      return FunctionDeclarationKind.BV_ROTATE_RIGHT;
+    } else if (kind.equals(Kind.BV_ROLI)) {
+      return FunctionDeclarationKind.BV_ROTATE_LEFT_BY_INT;
+    } else if (kind.equals(Kind.BV_RORI)) {
+      return FunctionDeclarationKind.BV_ROTATE_RIGHT_BY_INT;
     }
-    throw new UnsupportedOperationException("Can not discern formula kind " + kind);
+    return FunctionDeclarationKind.OTHER;
```
[the remark this answers is not in the document]
Comment (kfriedberger) [reply src/org/sosy_lab/java_smt/solvers/bitwuzla/BitwuzlaFormulaCreator.java:317]: There might be solver-specific extensions that JavaSMT does not want to support, or can not support. We introduced OTHER for such cases long time ago. 

This change also makes Bitwuzla more similar to the behaviour for other solvers in JavaSMT.

File: src/org/sosy_lab/java_smt/test/RotationVisitorTest.java [hunk 3]
```diff
@@ -0,0 +1,221 @@
...
+                FunctionDeclarationKind.BV_OR,
+                FunctionDeclarationKind.BV_SHL,
+                FunctionDeclarationKind.ITE,
+                FunctionDeclarationKind.ITE,
+                FunctionDeclarationKind.EQ,
+                FunctionDeclarationKind.EQ,
+                FunctionDeclarationKind.BV_SMOD,
+                FunctionDeclarationKind.BV_SMOD,
+                FunctionDeclarationKind.BV_LSHR,
+                FunctionDeclarationKind.BV_SUB);
```
[the remark this answers is not in the document]
Comment (kfriedberger) [reply src/org/sosy_lab/java_smt/test/RotationVisitorTest.java:57]: Yes, as soon as the solver provides a simpler way, we will change the test.

File: src/org/sosy_lab/java_smt/test/RotationVisitorTest.java [hunk 4]
```diff
@@ -0,0 +1,221 @@
...
+                FunctionDeclarationKind.BV_EXTRACT,
+                FunctionDeclarationKind.BV_EXTRACT,
+                FunctionDeclarationKind.BV_EXTRACT);
+        break;
+      default:
+        assertThat(functions).containsExactly(FunctionDeclarationKind.BV_ROTATE_LEFT);
+    }
+
+    // TODO dumpFormula crashes with Princess
+    // System.out.println(mgr.dumpFormula(bvmgr.equal(a, f)));
```
[the remark this answers is not in the document]
Comment (kfriedberger) [reply src/org/sosy_lab/java_smt/test/RotationVisitorTest.java:95]: Yes, that would be perfect.

**Pull request merged**